### ISSUE TITLE
iput progress output with client-side checksumming

### ISSUE THREAD

**Issue by Xophmeister** [irods/irods#5241]
## Bug Report

iRODS 4.2.7
Ubuntu 18.04.5 LTS

### What did you try to do?

    iput -PK large_file /zone/collection/

### Expected behavior

Per the documentation:

    -P  output the progress of the upload
    -K  verify checksum - calculate and verify the checksum on the data, both client-side and server-side, and store it in the catalog

### Observed behavior (including steps to reproduce, if applicable)

At regular intervals -- either time-based or size-based -- a log line should be output, informing how much has been uploaded, throughout the process. This works as expected when the `-K` option is _not_ supplied.

However, with `-K` -- presuming  uploading and client-side checksumming are done independently, rather than simultaneously -- no progress is logged during the upload. It silently waits until the upload is complete, then shows the progress logs quite quickly (i.e., more in keeping with local disk rates than less-stable network rates), during the client-side checksumming.

This is only perceptible when uploading very large files. At 500MiB, for example, our disk and network make this undetectable. However, at the ~10GiB scale, it's more obvious.

**Comment by trel** [comment]
This seems... to work for me?

```
$ ienv | grep version
irods_version - 4.2.8
```

make a 5G file
```
$ truncate -s5G 5gfile
```

`iput` with just progress... takes ~3 seconds
```
$ time iput -P 5gfile 5gfilep                                                                
0/1 -  0.00% of files done   0.000/5120.000 MB -  0.00% of file sizes done
Processing 5gfile - 5120.000 MB   2020-11-12.12:29:58
5gfile - 1960.000/5120.000 MB - 38.28% done   2020-11-12.12:29:59
5gfile - 2040.000/5120.000 MB - 39.84% done   2020-11-12.12:29:59
5gfile - 2160.000/5120.000 MB - 42.19% done   2020-11-12.12:29:59
5gfile - 2240.000/5120.000 MB - 43.75% done   2020-11-12.12:29:59
5gfile - 2320.000/5120.000 MB - 45.31% done   2020-11-12.12:29:59
5gfile - 2360.000/5120.000 MB - 46.09% done   2020-11-12.12:29:59
5gfile - 2480.000/5120.000 MB - 48.44% done   2020-11-12.12:29:59
5gfile - 2640.000/5120.000 MB - 51.56% done   2020-11-12.12:29:59
5gfile - 5120.000/5120.000 MB - 100.00% done   2020-11-12.12:30:01

real    0m3.386s
user    0m0.100s
sys     0m2.534s
```

`iput` with progress and server-side checksum takes 3 + ~18 = 21 seconds
```
$ time iput -Pk 5gfile 5gfilepk
0/1 -  0.00% of files done   0.000/5120.000 MB -  0.00% of file sizes done
Processing 5gfile - 5120.000 MB   2020-11-12.12:35:46
5gfile - 2040.000/5120.000 MB - 39.84% done   2020-11-12.12:35:47
5gfile - 2200.000/5120.000 MB - 42.97% done   2020-11-12.12:35:47
5gfile - 2320.000/5120.000 MB - 45.31% done   2020-11-12.12:35:47
5gfile - 2400.000/5120.000 MB - 46.88% done   2020-11-12.12:35:47
5gfile - 2480.000/5120.000 MB - 48.44% done   2020-11-12.12:35:47
5gfile - 2560.000/5120.000 MB - 50.00% done   2020-11-12.12:35:47
5gfile - 2640.000/5120.000 MB - 51.56% done   2020-11-12.12:35:47
5gfile - 2720.000/5120.000 MB - 53.12% done   2020-11-12.12:35:47
5gfile - 5120.000/5120.000 MB - 100.00% done   2020-11-12.12:35:49

real    0m21.563s
user    0m0.111s
sys     0m2.555s
```

`iput` with progress and both-sides checksum takes... ~18 + ~3 + ~18 = ~40-41 seconds
```
$ time iput -PK 5gfile 5gfilepk                                                              
0/1 -  0.00% of files done   0.000/5120.000 MB -  0.00% of file sizes done
Processing 5gfile - 5120.000 MB   2020-11-12.12:30:09
5gfile - 120.000/5120.000 MB -  2.34% done   2020-11-12.12:30:28
5gfile - 2480.000/5120.000 MB - 48.44% done   2020-11-12.12:30:29
5gfile - 2560.000/5120.000 MB - 50.00% done   2020-11-12.12:30:29
5gfile - 2680.000/5120.000 MB - 52.34% done   2020-11-12.12:30:29
5gfile - 2800.000/5120.000 MB - 54.69% done   2020-11-12.12:30:29
5gfile - 4040.000/5120.000 MB - 78.91% done   2020-11-12.12:30:30
5gfile - 4160.000/5120.000 MB - 81.25% done   2020-11-12.12:30:30
5gfile - 4600.000/5120.000 MB - 89.84% done   2020-11-12.12:30:30
5gfile - 5120.000/5120.000 MB - 100.00% done   2020-11-12.12:30:31

real    0m40.080s
user    0m17.749s
sys     0m3.397s
```

**Comment by Xophmeister** [comment]
I guess the confusion is that there's an apparent pause between launching the process and the progress logs showing up. Taking your last example:

```
$ time iput -PK 5gfile 5gfilePK                                                              
0/1 -  0.00% of files done   0.000/5120.000 MB -  0.00% of file sizes done
Processing 5gfile - 5120.000 MB   2020-11-12.12:30:09
5gfile - 120.000/5120.000 MB -  2.34% done   2020-11-12.12:30:28
5gfile - 2480.000/5120.000 MB - 48.44% done   2020-11-12.12:30:29
5gfile - 2560.000/5120.000 MB - 50.00% done   2020-11-12.12:30:29
5gfile - 2680.000/5120.000 MB - 52.34% done   2020-11-12.12:30:29
5gfile - 2800.000/5120.000 MB - 54.69% done   2020-11-12.12:30:29
5gfile - 4040.000/5120.000 MB - 78.91% done   2020-11-12.12:30:30
5gfile - 4160.000/5120.000 MB - 81.25% done   2020-11-12.12:30:30
5gfile - 4600.000/5120.000 MB - 89.84% done   2020-11-12.12:30:30
5gfile - 5120.000/5120.000 MB - 100.00% done   2020-11-12.12:30:31

real    0m40.080s
user    0m17.749s
sys     0m3.397s
```

Nearly 20 seconds elapse between the start and the first progress output. Maybe my presumption about the uploading being done in two phases (upload, then client-side checksum) is incorrect, but there's no output so you don't know. What's it doing in those 20 seconds (besides the 2.34% upload, which would have taken ~0.1s)? I noticed this because I was uploading ~10GiB files and `iput` was sitting there, outputting nothing, for tens of minutes. Fortunately I was patient enough, but I was beginning to worry!

**Comment by trel** [comment]
Steps...
1) 20sec - calculating the checksum on the client side (and outputting nothing...), then
2) 3 sec - sending the data (along with the calculated checksum) (with progress indicators), then
3) 20 sec - the server calculates the received data's checksum and compares it against the sent checksum (quietly again).

**Comment by kript** [comment]
We’ve had a few questions like this actually. Perhaps have -P print the checksum in conjunction with -V? Have it state it’s generating  the checksum?

**Comment by trel** [comment]
makes sense.

**Comment by trel** [comment]
perhaps something like...

"Calculating client checksum... [checksumvalue]"
progress...
progress...
progress...
progress...
"Calculating server checksum... [checksumvalue]"

**Comment by Xophmeister** [comment]
Yeah, this is more a UX thing than an actual software bug.

However, it's interesting that this is done in three distinct phases (client checksum, upload, server checksum). It's the same file stream: why not do it all at once? The client can send and checksum simultaneously, while the server can receive and checksum simultaneously. Checksumming is CPU bound, while the transfer is IO bound, so it wouldn't cause performance degradation during the upload and would result in a net speed up, which would be significant over large files.

**Comment by trel** [comment]
I think to your last point... we're limited by the current iRODS protocol.  The request to put data comes 'with' the calculated checksum... so it has to be calculated first, before it can be included.

**Comment by Xophmeister** [comment]
Fair enough; thanks, @trel :) Would it be worthwhile opening an enhancement request for this? I imagine protocol changes require a fair amount of effort!

**Comment by trel** [comment]
This issue has been marked as enhancement - for the better logging/output.

You are correct - changing the protocol is hard/cumbersome enough that we've promised not to do it for the lifetime of the 4.x series (because of backwards compatibility promises to 3.x).   It's always good to have requests logged so we don't forget them (and it more correctly captures the sentiment of the community) - please file one.   We'll categorize accordingly :)

**Comment by Xophmeister** [comment]
> This issue has been marked as enhancement - for the better logging/output.

Sorry, I meant the protocol change as an enhancement request. It's totally reasonable that this ticket, too, is an enhancement. Anyway, no worries: I will log a new issue for a protocol change proposal. Thanks again :)

**Comment by SwooshyCueb** [comment]
I'm in favor of pushing this to 4.2.10 or the 4.2 backlog if nobody else thinks they can get to it before the 4.2.9 deadline.

**Comment by trel** [comment]
Agreed.   Bumping.
